step-release-vis/src/app/services/environment.ts

```diff
@@ -14,21 +15,159 @@ export class EnvironmentService {
   // xs: 0-100, ys: timestamps
   getPolygons(jsonFile: string): Polygon[] {
     const environments = this.readJson(jsonFile);
-    return [];
+
+    var polys: Polygon[] = [];
+
+    // Assume there is only one environment
+    // Assume for now a candidate appears only once per environment (solution: give the polygons ids)
+
+    for (var environment of environments) {
+      var newPointSet: PointSet = new PointSet();
+      var lastPointSet: PointSet = new PointSet();
+      var lowerBounds : Map<string, Point[]> = new Map(); // both upper and lower bounds will contain the leftmost / rightmost point
+      var upperBounds : Map<string, Point[]> = new Map();
+      var lastTimeStamp = 0;
+
+      for (var snapshot of environment.snapshots) {
+        var update: [PointSet, number] = this.insertAndAdjustPoints(snapshot.cands_info, lastPointSet);
+        newPointSet = update[0];
+        this.addSnapshotToPolygons(lowerBounds, upperBounds, newPointSet.snapshot, update[1], snapshot.timestamp, lastTimeStamp);
+
+        lastPointSet = this.closePolygons(polys, lowerBounds, upperBounds, newPointSet); // delete inexisting ones
+        lastTimeStamp = snapshot.timestamp;
+      }
+
+      // draw the vertical line and add remaining polys
+      for (var i = 0; i < lastPointSet.snapshot.length; i ++) {
+        var name: string = lastPointSet.snapshot[i].candName;
+        this.addToMap(lowerBounds, name, {x: lastTimeStamp, y: lastPointSet.snapshot[i].position});
+        polys.push(this.createPolygon(lowerBounds.get(name), upperBounds.get(name)));
+      }
+    }
+
+    this.printPolys(polys);
+    return polys;
   }
 
-  private extractPoints(candsInfo: CandidateInfo[]): CandidatesSnapshot {
-    return [];
+  private printPolys (v : Polygon[]) {
+    console.log("These are the final polygons:");
+    for (var x of v) {
+      console.log("-----");
+      for (var point of x.points) {
+        console.log(point.x, point.y);
+      }
+    }
   }
 
-  // TODO(ancar): calculate the percentages for cands at given timestamp
-  private getPercentages(candsInfo: CandidateInfo[]): CandidatesSnapshot {
-    return [];
+  private addToMap (map: Map<string, Point[]>, key: string, point: Point) {
```

Comment:
there is a lot of code here, pls use what you learned in TDD, and make sure, it also works ;)